File: revisiting_ctci/chapter3/StackOfStacks_3_3.go
/*

have a SetOfStacks function that will make a struct holding stacks,
when a given stack reaches a certain size, close it and then build a new
one so that they all stay small.

when popping stacks, need to close stacks when they become empty


*/

package main

import (
	"./stack"
	"fmt"
)

// a stack of stacks, where each stack will have a max of 10 items
type SetOfStacks []stack.Stack

func (s SetOfStacks) String() string {
	outstring := ""
	for i, dat := range s {
		outstring = fmt.Sprintf("%v%v\t%v\n", outstring, i, dat)
	}
	return outstring
}

func (s *SetOfStacks) Push(d int) {
	if len(*s) == 0 || len((*s)[len(*s)-1].Ord) >= 10 {
		new_stack := stack.Stack{}
		new_stack.Push(d)
		*s = append((*s), new_stack)
	} else {
		(*s)[len(*s)-1].Push(d)
	}

}

func (s *SetOfStacks) Pop() int {
	out := (*s)[len(*s)-1].Pop()

	// if the last stack is empty then close it
	if len((*s)[len(*s)-1].Ord) == 0 {
		*s = (*s)[0 : len(*s)-1]
	}
	return out
}

func (s *SetOfStacks) Peek() int {
	return (*s)[len(*s)-1].Peek()
}

func main() {

	test := SetOfStacks{}

	for i := 0; i < 22; i++ {
		test.Push(i)
	}

	fmt.Println(test)

	for i := 0; i < 3; i++ {
		x := test.Pop()
		fmt.Println(x)
	}

	fmt.Println(test)

}

File: revisiting_ctci/chapter3/StackOfStacks_3_3.py
from stack import Stack

class SetOfStacks():
	def __init__(self, max_len):
		self.__all = [Stack()]
		self.__max_len = max_len

	def Push(self, data):
		if len(self.__all[-1].Data) >= self.__max_len:
			new = Stack()
			new.Push(data)
			self.__all.append(new)
		else:
			self.__all[-1].Push(data)

	def Pop(self):
		out = self.__all[-1].Pop()
		if len(self.__all[-1].Data) == 0:
			self.__all = self.__all[:-1]
		return out

	def Peek(self):
		return self.__all[-1].Peek()

	def __repr__(self):
		outstring = ''
		for i, data in enumerate(self.__all):
			line = f'{i}:\t{data}\n'
			outstring += line
		return outstring


if __name__ == '__main__':

	test = SetOfStacks(5)

	for i in range(17):
		test.Push(i)

	print(test)
	
	print('Popping:\n')
	for i in range(3):
		print(test.Pop())

	print('\nFinal:\n')
	print(test)